### DISCUSSION TITLE
envd `native local` mode

### DISCUSSION THREAD

**Discussion by gaocegege** [tensorchord/envd#476] [Ideas]
Now we run the docker container and use ssh or jupyter to attach into the container. I am wondering if we can, or should, support the native local mode:

We use virtual env or conda to create the environment in this mode.

The benefit is that we can provide one manifest, and users can use envd to create an environment on host OS, on docker, and on kubernetes in the future.

**Comment by gaocegege** [comment]
The user experience would like:

```bash
# Use the conda instead of docker to create the environment
$ envd --conext local-conda up
install python packages and conda packages...
configure environment variables...
# The shell is configured with `conda activate env`
(env) $ 
```

**Reply by gaocegege** [reply]
cc @popfido 

I'd appreciate it if you give some feedback.

**Reply by popfido** [reply]
Does this local mode idea come from [starwhale](https://github.com/star-whale/starwhale)?

**Reply by gaocegege** [reply]
Nope, just get the idea from my dream last night.

**Comment by popfido** [comment]
I think it's okay. 

But there exists several concerns that I may take care of:

- The complexity. Ppl use/know `envd` might already had a local env. Thus when supporting `local` mode, one thing I think that needs to be concerned is how to deal with various situation when there already has local/existence `conda` envs. Compared with its complexity when dealing those synchronization. I would prefer supporting `on kubernetes` model a higher priority than supporting `local` mode.
- Besides, the reason why developers would prefer docker/k8s is that configuration in various local environment is so messy. Going back to support `local` mode with treating `envd` as an extra wrapper on local `conda` seems quite redundant:
  - Developers who have experience handling environment would have no needs on `local` mode.
  - Developers who have troubles handling environment or need a unified consistent environment would directly choose to use `docker` mode or `k8s` mode. Without any concern about setting `local` env.

The only group of developers/researchers I think that need `local` mode is defined as:

- They require unified consistent environments.
- They also require local environments.
- They have not yet installed `conda` or other local environment management tools.

If so, I would further suggest that the `conda` protocol should be hide completely when using`envd` to manage the local envs to reduce complexity of `local` mode when dealing various compatibility problems with other local env.

Just my personal opinion.
